martin/build.rs

```diff
@@ -1,13 +1,115 @@
-fn main() -> std::io::Result<()> {
-    #[cfg(feature = "webui")]
-    {
-        static_files::NpmBuild::new("martin-ui")
-            .install()?
-            .run("build")?
-            .target("martin-ui/dist")
-            .change_detection()
-            .to_resource_dir()
-            .build()?;
+#[cfg(feature = "webui")]
+use std::fs;
+#[cfg(feature = "webui")]
+use std::path::Path;
+
+#[cfg(feature = "webui")]
+fn copy_file_tree(src: &Path, dst: &Path, exclude_dirs: &[&str]) {
+    assert!(
+        src.is_dir(),
+        "source for the copy operation is not an existing directory"
+    );
+    let _ = fs::remove_dir_all(dst); // ignore if dir does not exist
+    fs::create_dir_all(dst).unwrap_or_else(|e| {
+        panic!(
+            "failed to create destination directory {}: {e}",
+            dst.display()
+        )
+    });
+    let excludes = exclude_dirs.iter().map(|v| src.join(v)).collect::<Vec<_>>();
+
+    let mut it = walkdir::WalkDir::new(src).into_iter();
+    loop {
+        let entry = match it.next() {
+            None => break,
+            Some(Err(err)) => panic!("failed to read directory entry: {err}"),
+            Some(Ok(entry)) => entry,
+        };
+        if excludes.iter().any(|v| v == entry.path()) {
+            it.skip_current_dir();
+            continue;
+        }
+
+        // Get the relative path of the entry
+        let dst_path = dst.join(
+            entry
+                .path()
+                .strip_prefix(src)
+                .expect("path is not a prefix of the source directory"),
+        );
+
+        // Tell Cargo to monitor all root level entries for changes
+        if dst_path.components().count() == 1 {
+            println!("cargo:rerun-if-changed={}", entry.path().display());
+        }
```

Comment:
Is this not already covered by `static_files::NpmBuild::change_detection` below, or am I missing something?

As far as I read this, we can keep either, but lets not keep duplicate code..

Reply:
good catch!